In [1]:
import pandas as pd

In [2]:
data=pd.read_csv('Spam_Ham/spam_ham_large_dataset.csv')

In [3]:
data

label                                            message
0     ham                          Let's go for lunch today.
1    spam  This is your final notice. Pay now to avoid pe...
2     ham                         Did you finish the report?
3     ham                          Let's go for lunch today.
4    spam  URGENT! Your account has been suspended. Verif...
..    ...                                                ...
195  spam  This is your final notice. Pay now to avoid pe...
196  spam  URGENT! Your account has been suspended. Verif...
197   ham                           Let's catch up tomorrow.
198   ham               Don't forget to bring the documents.
199  spam     Your loan is approved. Contact us immediately.

[200 rows x 2 columns]

In [8]:
import re
from nltk.corpus import stopwords

In [5]:
from nltk.stem import WordNetLemmatizer
lemma=WordNetLemmatizer()

In [16]:
corpus=[]
for i in range(len(data)):
    rem_spl=re.sub('[^A-Za-z]',' ',data['message'][i])
    low_sent=rem_spl.lower()
    split_sent=low_sent.split()
    #print(split_sent)
    final=[lemma.lemmatize(word,pos='v') for word in split_sent if word not in stopwords.words('english')]
    final=" ".join(final)
    corpus.append(final)

In [17]:
corpus

['let go lunch today',
 'final notice pay avoid penalty',
 'finish report',
 'let go lunch today',
 'urgent account suspend verify immediately',
 'hey still meet today',
 'reschedule meet',
 'urgent account suspend verify immediately',
 'hey still meet today',
 'limit time offer buy get',
 'send file shortly',
 'final notice pay avoid penalty',
 'finish report',
 'minutes',
 'win free iphone today click link',
 'please call free',
 'get rich fast join program',
 'win free iphone today click link',
 'win free iphone today click link',
 'final notice pay avoid penalty',
 'congratulations free gift voucher click link claim',
 'finish report',
 'congratulations free gift voucher click link claim',
 'urgent account suspend verify immediately',
 'urgent account suspend verify immediately',
 'please call free',
 'exclusive deal act',
 'let catch tomorrow',
 'forget bring document',
 'forget bring document',
 'win free iphone today click link',
 'forget bring document',
 'reschedule meet',
 'p

Unigram

In [19]:
from sklearn.feature_extraction.text import CountVectorizer
cv=CountVectorizer(max_features=50,binary=True,ngram_range=(1,1))

In [ ]:
cv.fit_transform(corpus)

<200x50 sparse matrix of type '<class 'numpy.int64'>'
	with 689 stored elements in Compressed Sparse Row format>

| Part                    | Meaning                         |
| ----------------------- | ------------------------------- |
| **200x50**              | 200 messages × 50 features      |
| **sparse matrix**       | only non-zero values stored     |
| **int64**               | integer counts                  |
| **689 stored elements** | non-zero entries                |
| **CSR format**          | compressed sparse row structure |


In [21]:
X=cv.fit_transform(corpus).toarray()

In [23]:
cv.vocabulary_

{'let': 23,
 'today': 42,
 'final': 16,
 'notice': 27,
 'pay': 28,
 'avoid': 3,
 'penalty': 29,
 'urgent': 44,
 'account': 0,
 'suspend': 41,
 'verify': 47,
 'immediately': 20,
 'meet': 26,
 'reschedule': 36,
 'get': 19,
 'send': 39,
 'file': 15,
 'shortly': 40,
 'win': 48,
 'free': 18,
 'iphone': 21,
 'click': 9,
 'link': 24,
 'please': 30,
 'call': 5,
 'rich': 37,
 'fast': 14,
 'join': 22,
 'program': 34,
 'claim': 8,
 'exclusive': 13,
 'deal': 11,
 'act': 1,
 'catch': 7,
 'tomorrow': 43,
 'forget': 17,
 'bring': 4,
 'document': 12,
 'select': 38,
 'cash': 6,
 'prize': 33,
 'reply': 35,
 'yes': 49,
 'loan': 25,
 'approve': 2,
 'contact': 10,
 'us': 45,
 'vacation': 46,
 'postpone': 32,
 'pm': 31}

Bigram

In [29]:
cv1=CountVectorizer(max_features=100,binary=True,ngram_range=(1,2))

In [30]:
cv1.fit_transform(corpus)

<200x100 sparse matrix of type '<class 'numpy.int64'>'
	with 1228 stored elements in Compressed Sparse Row format>

In [31]:
cv1.vocabulary_

{'let': 46,
 'lunch': 54,
 'today': 85,
 'let go': 48,
 'go lunch': 40,
 'final': 31,
 'notice': 57,
 'pay': 59,
 'avoid': 5,
 'penalty': 61,
 'final notice': 32,
 'notice pay': 58,
 'pay avoid': 60,
 'avoid penalty': 6,
 'urgent': 88,
 'account': 0,
 'suspend': 81,
 'verify': 94,
 'immediately': 41,
 'urgent account': 89,
 'account suspend': 1,
 'suspend verify': 82,
 'verify immediately': 95,
 'meet': 55,
 'reschedule': 72,
 'reschedule meet': 73,
 'limit': 49,
 'time': 83,
 'buy': 9,
 'get': 38,
 'limit time': 50,
 'time offer': 84,
 'send': 78,
 'file': 29,
 'shortly': 80,
 'send file': 79,
 'file shortly': 30,
 'win': 96,
 'free': 35,
 'iphone': 42,
 'click': 18,
 'link': 51,
 'win free': 97,
 'free iphone': 36,
 'iphone today': 43,
 'today click': 86,
 'click link': 19,
 'please': 62,
 'call': 10,
 'please call': 63,
 'call free': 11,
 'rich': 74,
 'fast': 27,
 'join': 44,
 'program': 69,
 'get rich': 39,
 'rich fast': 75,
 'fast join': 28,
 'join program': 45,
 'claim': 16,
 'ex

Trigram

In [38]:
cv2=CountVectorizer(max_features=500,binary=True,ngram_range=(1,3))

In [39]:
cv2.fit_transform(corpus)

<200x162 sparse matrix of type '<class 'numpy.int64'>'
	with 1790 stored elements in Compressed Sparse Row format>

In [40]:
cv2.vocabulary_

{'let': 76,
 'go': 64,
 'lunch': 89,
 'today': 141,
 'let go': 79,
 'go lunch': 65,
 'lunch today': 90,
 'let go lunch': 80,
 'go lunch today': 66,
 'final': 43,
 'notice': 96,
 'pay': 102,
 'avoid': 7,
 'penalty': 105,
 'final notice': 44,
 'notice pay': 97,
 'pay avoid': 103,
 'avoid penalty': 8,
 'final notice pay': 45,
 'notice pay avoid': 98,
 'pay avoid penalty': 104,
 'finish': 46,
 'report': 119,
 'finish report': 47,
 'urgent': 145,
 'account': 0,
 'suspend': 135,
 'verify': 152,
 'immediately': 70,
 'urgent account': 146,
 'account suspend': 1,
 'suspend verify': 136,
 'verify immediately': 153,
 'urgent account suspend': 147,
 'account suspend verify': 2,
 'suspend verify immediately': 137,
 'hey': 67,
 'still': 132,
 'meet': 91,
 'hey still': 68,
 'still meet': 133,
 'meet today': 94,
 'hey still meet': 69,
 'still meet today': 134,
 'reschedule': 120,
 'reschedule meet': 121,
 'limit': 81,
 'time': 138,
 'offer': 99,
 'buy': 11,
 'get': 58,
 'limit time': 82,
 'time offer'

In [41]:
X1=cv2.fit_transform(corpus).toarray()

In [42]:
X1

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]], dtype=int64)

N-gram is used to differentiate the vectors while training the model. So model will differentiate the vocabulary well in the form of vectors